time/src/time.cpp

```diff
@@ -0,0 +1,559 @@
+
+    // fix os-specific format string difference between macos/windows and linux
+    lua_pushfstring(L, "%lld.%09d", static_cast<long long int>(timespec.tv_sec), timespec.tv_nsec);
+    return 1;
+}
+
+static int duration__add(lua_State* L)
+{
+    uv_timespec64_t left = getTimespecFromDuration(L, 1);
+    uv_timespec64_t right = getTimespecFromDuration(L, 2);
+
+    uv_timespec64_t result = {left.tv_sec + right.tv_sec, left.tv_nsec + right.tv_nsec};
+    if (result.tv_nsec > NANOSECONDS_PER_SECOND)
+    {
+        result.tv_sec += 1;
+        result.tv_nsec -= NANOSECONDS_PER_SECOND;
+    }
+
+    return createDurationFromTimespec(L, result);
+}
+
+static int duration__sub(lua_State* L)
+{
+    uv_timespec64_t left = getTimespecFromDuration(L, 1);
+    uv_timespec64_t right = getTimespecFromDuration(L, 2);
+
+    uv_timespec64_t result = {left.tv_sec - right.tv_sec, left.tv_nsec - right.tv_nsec};
+    if (result.tv_nsec < 0)
+    {
+        result.tv_sec -= 1;
+        result.tv_nsec += NANOSECONDS_PER_SECOND;
+    }
+
+    return createDurationFromTimespec(L, {result.tv_sec >= 0 ? result.tv_sec : 0, result.tv_nsec >= 0 ? result.tv_nsec : 0});
+}
+
+static int duration__eq(lua_State* L)
+{
+    uv_timespec64_t left = getTimespecFromDuration(L, 1);
+    uv_timespec64_t right = getTimespecFromDuration(L, 2);
+
+    lua_pushboolean(L, left.tv_sec == right.tv_sec && left.tv_nsec == right.tv_nsec);
+    return 1;
+}
+
+static int duration__lt(lua_State* L)
+{
+    uv_timespec64_t left = getTimespecFromDuration(L, 1);
+    uv_timespec64_t right = getTimespecFromDuration(L, 2);
+
+    lua_pushboolean(L, left.tv_sec < right.tv_sec || (left.tv_sec == right.tv_sec && left.tv_nsec < right.tv_nsec));
+    return 1;
+}
+
+static int duration__le(lua_State* L)
+{
+    uv_timespec64_t left = getTimespecFromDuration(L, 1);
+    uv_timespec64_t right = getTimespecFromDuration(L, 2);
+
+    lua_pushboolean(L, left.tv_sec < right.tv_sec || (left.tv_sec == right.tv_sec && left.tv_nsec <= right.tv_nsec));
+    return 1;
+}
+
+// Instants
+static uv_timespec64_t getTimespecFromInstant(lua_State* L, int idx)
+{
+    return *static_cast<uv_timespec64_t*>(luaL_checkudata(L, idx, kInstantType));
+}
+
+// Methods
+static int instant_elapsed(lua_State* L)
+{
+    lua_pushnumber(L, sinceTimespec(getTimespecFromInstant(L, 1)));
+    return 1;
+}
+
+// Metamethods
+static int instant__sub(lua_State* L)
+{
+    uv_timespec64_t left = getTimespecFromInstant(L, 1);
+    uv_timespec64_t right = getTimespecFromInstant(L, 2);
+
+    return createDurationFromSeconds(L, static_cast<double>(diffTimespecs(left, right)));
+}
+
+namespace duration
+{
+int lua_nanoseconds(lua_State* L)
+{
+    // int32_t doesn't have enough precision for nanoseconds
+    int64_t nanoseconds = static_cast<int64_t>(luaL_checknumber(L, 1));
+    if (nanoseconds < 0)
+    {
+        luaL_error(L, "duration cannot be negative");
+        return 0;
+    }
```

Comment:
```suggestion
        luaL_error(L, "duration cannot be negative");
```

`luaL_error` is noret.